Add project settings for default importers

mlm-games opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

Closes https://github.com/godotengine/godot-proposals/issues/12770.
Closes #108512 

Doesn't import older files by default to maintain .import settings, could use the tools -> upgrade project files for that

<img width="1174" height="527" alt="image" src="https://github.com/user-attachments/assets/10695790-32c9-4b05-8929-2145cb27c38c" />

Toggling it on and adding a svg file would import it as a SVGTexture else as a CompressedTexture2D

**Production edit:** Closes https://github.com/godotengine/godot-proposals/issues/2500

Ivorforce commented:
> Only the mono linux editor says documentation is incomplete? Is it just a pipline problem because others don't seem cancelled..

Some tests just the build process, some do unit tests, and some do additional stuff like documentation checks. I think only the Linux Editor w/ Mono target checks the documentation.

mlm-games commented:
> Edit: My guess is, your `GLOBAL_DEF` is called in the wrong location. The docs seem to complain that your added setting does not exist, so it's not registered at the time it's checking.

Thanks for the ref, moved it to ProjectSettings (and removed some unneeded flags i used looking at others there, hadn't put it there initially because the filesystem/import settings were not present initially)

mlm-games commented:
FPed the changed (changed the names of some due to it being confusing while handling importer-settings (named as settings initially) and default-importer-settings)

mlm-games commented:
<img width="1209" height="733" alt="image" src="https://github.com/user-attachments/assets/e718399e-3ce9-44a7-9143-5897b88c4ab9" />

Made the changes (don't mind the tab names)

KoBeWi commented on core/io/resource_importer.cpp:
```diff
@@ -508,6 +508,13 @@ void ResourceFormatImporter::get_importers(List<Ref<ResourceImporter>> *r_import
 }
 
 Ref<ResourceImporter> ResourceFormatImporter::get_importer_by_file(const String &p_file) const {
+	String extension = p_file.get_extension().to_lower();
```
```suggestion
	const String extension = p_file.get_extension().to_lower();
```

KoBeWi commented on core/io/resource_importer.cpp:
```diff
@@ -526,6 +533,25 @@ Ref<ResourceImporter> ResourceFormatImporter::get_importer_by_file(const String
+	if (!ProjectSettings::get_singleton()->has_setting("default_resource_importers/" + p_extension)) {
+		return Ref<ResourceImporter>();
+	}
+	const String importer_name = (String)GLOBAL_GET("default_resource_importers/" + p_extension);
```
```suggestion
	const String importer_name = GLOBAL_GET("default_resource_importers/" + p_extension);
```

KoBeWi commented on editor/import/import_defaults_editor.cpp:
```diff
@@ -77,40 +80,166 @@ class ImportDefaultsEditorSettings : public Object {
 void ImportDefaultsEditor::_notification(int p_what) {
 	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE:
+		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
```
Needs `check_changed_settings_in_group()` call.

KoBeWi commented on editor/import/import_defaults_editor.h:
```diff
@@ -33,32 +33,41 @@
 	Button *save_defaults = nullptr;
 	Button *reset_defaults = nullptr;
 
 	EditorInspector *inspector = nullptr;
```
Unused.

mlm-games replied:
It is used for adding the second tab 

```c++
tabs->add_child(default_importers_tab);
```

KoBeWi replied:
I meant the `inspector`.
Sorry, GitHub can't handle single-line comments.

KoBeWi commented on editor/import/import_defaults_editor.h:
```diff
@@ -33,32 +33,41 @@
+	EditorInspector *extension_importer_inspector = nullptr;
+
+	ImporterDefaultSettings *importer_settings = nullptr;
 
```
Unnecessary empty line.

KoBeWi commented on editor/import/import_defaults_editor.cpp:
```diff
@@ -195,11 +327,43 @@ void ImportDefaultsEditor::clear() {
 			_update_importer();
 		}
 	}
+	// Only triggers if anything is selected
```
```suggestion
	// Only triggers if anything is selected.
```
Same in other comments.

KoBeWi commented on editor/import/import_defaults_editor.cpp:
```diff
@@ -77,40 +80,167 @@ class ImportDefaultsEditorSettings : public Object {
 	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE:
+		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
+			EditorSettings::get_singleton()->check_changed_settings_in_group("importer_defaults/default_setting_importers");
```
Not like this 🙃
Closest example would be
https://github.com/godotengine/godot/blob/2d113cc224cb9be07866d003819fcef2226a52ea/editor/animation/animation_track_editor.cpp#L1475-L1484

mlm-games replied:
> Although I realized that this is not needed, because EditorInspector handles it automatically.

Oop, removed it then

AThousandShips commented on editor/import/import_defaults_editor.cpp:
```diff
@@ -77,40 +81,158 @@ class ImportDefaultsEditorSettings : public Object {
+	bool save_extension_importer(const String &p_extension) const {
+		Variant index;
+		if (_get(p_extension, index)) {
+			Variant value; // Fallback to default (unset the setting).
```
You undid this change

mlm-games replied:
Added it this time

mlm-games commented:
> It's the button that says "add suggestion to batch"

Oh probably didn't notice it 😅

KoBeWi commented on editor/import/import_defaults_editor.cpp:
```diff
@@ -211,25 +355,38 @@ ImportDefaultsEditor::ImportDefaultsEditor() {
+	extension_importers_label->set_text(TTRC("Select default importer for each file extension:"));
+	default_importers_tab->add_child(extension_importers_label);
+
+	extension_importer_inspector = memnew(EditorInspector);
+	default_importers_tab->add_child(extension_importer_inspector);
+	extension_importer_inspector->set_v_size_flags(SIZE_EXPAND_FILL);
```
```suggestion
	extension_importer_inspector = memnew(EditorInspector);
	extension_importer_inspector->set_v_size_flags(SIZE_EXPAND_FILL);
	extension_importer_inspector->set_property_name_style(EditorPropertyNameProcessor::STYLE_RAW);
	extension_importer_inspector->set_use_settings_name_style(false);
	default_importers_tab->add_child(extension_importer_inspector);
```

mlm-games committed:
Add project settings for default importers

asteinlein commented:
Is there anything than I can do to help move this forward? I'm extensively using SVGs, and just came across this PR now while searching for a way to make `DPITexture` default for SVGs in my project, after for the n-th time having forgot to change one stray SVG file. Would love for this to get merged, and I can help if need be.
